cli/command/trust/signer_add.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/docker/notary/client"
+	"github.com/docker/notary/passphrase"
+	store "github.com/docker/notary/storage"
+	"github.com/docker/notary/tuf/data"
+	"github.com/docker/notary/utils"
+	"github.com/spf13/cobra"
+)
+
+type signerAddOptions struct {
+	key    string
+	signer string
+	images opts.ListOpts
+}
+
+func newSignerAddCommand(dockerCli command.Cli) *cobra.Command {
+	var options signerAddOptions
+	cmd := &cobra.Command{
+		Use:   "signer-add [OPTIONS] NAME ",
+		Short: "Add a signer",
+		Args:  cli.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.signer = args[0]
+			return addSigner(dockerCli, &options)
+		},
+	}
+	flags := cmd.Flags()
+	flags.StringVarP(&options.key, "key", "k", "", "Path to the signer key to add")
+	options.images = opts.NewListOpts(nil)
+	flags.VarP(&options.images, "image", "i", "Images to add signer to")
+	return cmd
+}
+
+func addSigner(cli command.Cli, options *signerAddOptions) error {
+	signerName := options.signer
+
+	// TODO@ashfall: instead of failing if no key is provided, ask user if
+	// they want to generate a key
+	if options.key == "" {
+		return fmt.Errorf("path to a valid key must be provided")
+	}
+	if options.images.Len() < 1 {
+		return fmt.Errorf("at least one image must be provided")
+	}
+
+	for _, imageName := range options.images.GetAll() {
+		if err := addSignerToImage(cli, signerName, imageName, options.key); err != nil {
+			fmt.Fprintf(cli.Out(), "Failed to add signer to %s: %s\n", imageName, err)
+			continue
+		}
+		fmt.Fprintf(cli.Out(), "Successfully added signer: %s to %s\n", signerName, imageName)
+	}
+	return nil
+}
+
+func addSignerToImage(cli command.Cli, signerName string, imageName string, keyPath string) error {
+	_, ref, repoInfo, authConfig, err := getImageReferencesAndAuth(cli, imageName)
+	if err != nil {
+		return err
+	}
+
+	notaryRepo, err := trust.GetNotaryRepository(cli, repoInfo, *authConfig, "push", "pull")
+	if err != nil {
+		return trust.NotaryError(ref.Name(), err)
+	}
+
+	if err = notaryRepo.Update(false); err != nil {
+		switch err.(type) {
+		case client.ErrRepoNotInitialized, client.ErrRepositoryNotExist:
+			fmt.Fprintf(cli.Out(), "Initializing %s...\n", imageName)
+			if err := getOrGenerateRootKeyAndInitRepo(notaryRepo); err != nil {
+				return trust.NotaryError(ref.Name(), err)
+			}
+			fmt.Fprintf(cli.Out(), "Successfully initialized %q\n", notaryRepo.GetGUN().String())
+		default:
+			return trust.NotaryError(repoInfo.Name.Name(), err)
+		}
+	}
+
+	fmt.Fprintf(cli.Out(), "\nAdding signer \"%s\" to %s...\n", signerName, imageName)
+	newSignerRoleName := data.RoleName(path.Join(data.CanonicalTargetsRole.String(), signerName))
+
+	passRetriever := passphrase.PromptRetrieverWithInOut(cli.In(), cli.Out(), nil)
+	importers, err := getImporters(trust.GetTrustDirectory())
+	if err != nil {
+		return err
+	}
+	if err := importKeys(newSignerRoleName.String(), keyPath, passRetriever, importers); err != nil {
```

Comment:
from discussion: this might need to take `signerName` instead of the full `targets/<signerName>`